app/commands/solution/search_user_solutions.rb

```diff
@@ -21,59 +21,129 @@ def initialize(user, criteria: nil, track_slug: nil, status: nil, mentoring_stat
-      filter_track!
-      filter_status!
-      filter_mentoring_status!
-      sort!
-
-      @solutions.page(page).per(per)
+      results = Exercism.opensearch_client.search(index: "#{Rails.env}-solutions", body: search_body)
+
+      solution_ids = results["hits"]["hits"].map { |hit| hit["_source"]["id"] }
+      solutions = solution_ids.present? ?
+        Solution.where(id: solution_ids).
+          includes(:exercise, :track).
+          order(Arel.sql("FIND_IN_SET(id, '#{solution_ids.join(',')}')")).
+          to_a : []
+
+      total_count = results["hits"]["total"]["value"].to_i
+      Kaminari.paginate_array(solutions, total_count: total_count).
+        page(page).per(per)
+    rescue StandardError
+      Fallback.(user, page: page, per: per, track_slug: track_slug, status: status, mentoring_status: mentoring_status,
+criteria: criteria, order: order)
     end
 
     private
     attr_reader :user, :criteria, :track_slug, :status, :mentoring_status,
       :per, :page, :order,
       :solutions
 
-    def filter_criteria!
-      return if criteria.blank?
+    def search_body
+      {
+        query: search_query,
+        sort: search_sort,
 
-      @solutions = @solutions.joins(exercise: :track)
-      criteria.strip.split(" ").each do |crit|
-        @solutions = @solutions.where(
-          "exercises.title LIKE ? OR tracks.title LIKE ?",
-          "%#{crit}%",
-          "%#{crit}%"
-        )
-      end
+        # Only return the solution IDs, not the entire document, to improve performance
+        _source: [:id],
+
+        # Paging information
+        from: (page - 1) * per,
+        size: per
+      }
     end
 
-    def filter_track!
-      return if track_slug.blank?
+    def search_query
+      {
+        bool: {
+          must: [
+            { term: { 'user.id': user.id } },
+            track_slug.blank? ? nil : { term: { 'track.slug': track_slug } },
+            status.blank? ? nil : { term: { status: status } },
+            mentoring_status.blank? ? nil : { term: { mentoring_status: mentoring_status } },
+            criteria.blank? ? nil : { query_string: { query: "*#{criteria}*", fields: ['exercise.title', 'track.title'] } }
+          ].compact
+        }
+      }
+    end
 
-      @solutions = @solutions.joins(exercise: :track).
-        where('tracks.slug': track_slug)
+    def search_sort
+      [{ last_iterated_at: { order: order&.to_sym == :oldest_first ? :desc : :asc } }]
     end
 
-    def filter_status!
-      return if status.blank?
+    class Fallback
+      include Mandate
 
-      @solutions = @solutions.where(status: status)
-    end
+      def initialize(user, criteria: nil, track_slug: nil, status: nil, mentoring_status: nil, page: nil, per: nil, order: nil)
+        @user = user
+        @criteria = criteria
+        @track_slug = track_slug
+        @status = status
+        @mentoring_status = mentoring_status
+        @page = page
+        @per = per
+        @order = order
+      end
```

Comment:
I'd be pretty tempted to get rid of the kwargs and the attr_readers and just use initialize_with here, as its an internal class, rather than something more "public".